zebra-state/src/in_memory/chains.rs

```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
 use futures::prelude::*;
 use std::{
     pin::Pin,
+    sync::Arc,
     task::{Context, Poll},
 };
-use tower::Service;
+use tower::{Service, ServiceExt};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+/// A representation of the chain state at a given height
+#[derive(Clone, Debug)]
+struct ChainState {
+    block: Arc<Block>,
+    // TODO: add more state here
+}
+
+/// A persistent data structure representing the end of a chain
+struct VolatileChain(im::OrdMap<BlockHeight, ChainState>);
+
+impl VolatileChain {
+    fn contains(&self, height: BlockHeight) -> bool {
+        todo!()
+    }
+}
 
 /// A service wrapper that tracks multiple chains, handles reorgs, and persists
 /// blocks to disk once they're past the reorg limit
 pub(crate) struct ChainsState<S> {
     /// The inner state service that only tracks a single chain
     inner: S,
+    /// The set of chains
+    //
+    // might need to use a map type and pop / reinsert with cummulative work as the index
+    chains: Vec<VolatileChain>,
+}
+
+impl<S> ChainsState<S> {
+    fn insert(&mut self, block: impl Into<Arc<Block>>) -> Result<BlockHeaderHash, Error> {
+        let block = block.into();
+        let hash = block.hash();
+        let height = block.coinbase_height().unwrap();
+        let parent_height = BlockHeight(height.0 - 1);
+
+        for chain in self
+            .chains
+            .iter_mut()
+            .filter(|chain| chain.contains(parent_height))
+        {
+            let parent_state = chain
+                .0
+                .get(&parent_height)
+                .expect("block with one less height must exist");
+            let parent_hash = parent_state.block.hash();
+
+            if parent_hash != block.header.previous_block_hash {
+                continue;
+            }
+
+            let was_present = if chain.contains(height) {
```

Comment:
I think there's a really subtle bug here, triggered by another bug in the syncer.

1. The syncer submits a duplicate block for verification
2. It passes verification, based on the content of the parent block's chain state
3. The state already has 2 alternate tips, one contains the duplicate block, and one does not

If the tip that contains the duplicate block appears earlier in chains, the state layer panics.

If the tip that contains the parent block, but not the duplicate block, appears earlier in chains, the state layer creates a duplicate entry in chains. So now there are two entries in chains that contain exactly the same blocks.

Here's a different algorithm that ensures we don't end up with duplicate chains:
1. Store the ChainList as a HashMap by block hash. Each Chain in the ChainList is stored as an OrdMap by block height.
2. When the state gets a new block, check for its hash in the ChainList. If it exists, error or panic. (Probably error for now, until we fix the syncer.)
3. If the new block does not exist, check for its parent block in ChainList. If the parent not exist, error or panic. (Error until #737 is merged, then panic.)
4. Take a copy of the parent block's Chain, and insert the new block as a child of the parent block. Add the new Chain to the ChainList.